### PR TITLE
1358 import tn corpus

### PR DESCRIPTION
**Pull request by flooie**
TN Corpus Importer.

This PR handles the importing of roughly 1.5k cases from the TN Workers Comp Claims/Appeals.

### CODE REVIEW

File: cl/corpus_importer/management/commands/import_tn.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+            )
+            if x["court"] == court
+        ]
+        for key, group in itertools.groupby(tn, lambda x: x["pub_date"][:4]):
+            count = 1
+            for case in list(group):
+                case["neutral_citation"] = "%s %s %s" % (key, reporter, count)
+                results.append(case)
+                count += 1
+    return results
```
Comment (mlissner) [inline cl/corpus_importer/management/commands/import_tn.py:91]: This function does...a lot that isn't adding neutral citations. Seems like you should have one function that opens the data and returns it, and a second that loops over that data and adds neutral citations. 

I also find the logic here very strange:

 - for each of the two courts
 - loop through the entire data set to filter it
 - add the filtered data to a var if it matches the court name we're working on
 - sort and group the data and do neutral citations

Why not just loop through the data once to organize it (in one function called from `import_tn_corpus`, and then have this function just add neutral citations? Should be much clearer.

File: cl/corpus_importer/management/commands/import_tn.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+def make_item(case):
+    """Make an import item for our save everything function
+
+    :param case: Case data received from the court.
+    :return: Case information.
+    :type: dict
+    """
+
+    judges = case["judge"]
+    first_name = judges.split(" ")[0].title()
```
Comment (mlissner) [inline cl/corpus_importer/management/commands/import_tn.py:29]: It's probably fine here, but I almost never use Python's `title` util. We have one in Juriscraper that handles things much better with things like `MacDonald`, `L'Oreal`, etc. I think it's called titlecase().

File: cl/corpus_importer/management/commands/import_tn.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        positions__court_id=case["court"],
+        is_alias_of=None,
+        name_first=first_name,
+    )
+    # Four Panelists exist, one retired end of 2019 and the other joined.
+    panelists = case["judge"]
+    if case["court"] == "tennworkcompapp":
+        exclude = (
+            "Marshall"
+            if parser.parse(case["pub_date"]).date().year == 2020
```
Comment (mlissner) [inline cl/corpus_importer/management/commands/import_tn.py:41]: Parse the date once, put it in a variable, then use that here.

Comment (mlissner) [reply cl/corpus_importer/management/commands/import_tn.py:41]: Handy that they retired at the end of the year, wow.

File: cl/corpus_importer/management/commands/import_tn.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    judges = case["judge"]
+    first_name = judges.split(" ")[0].title()
+
+    lead_author = Person.objects.get(
+        positions__court_id=case["court"],
+        is_alias_of=None,
+        name_first=first_name,
+    )
+    # Four Panelists exist, one retired end of 2019 and the other joined.
+    panelists = case["judge"]
```
Comment (mlissner) [inline cl/corpus_importer/management/commands/import_tn.py:37]: I'd put this in the `else` to make it more clear.

File: cl/corpus_importer/management/commands/import_tn.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+            required=True,
+        )
+        parser.add_argument(
+            "--log",
+            action="store_true",
+            default=False,
+            help="Choose to view info log lines.",
+        )
+        parser.add_argument(
+            "--skip-until", default=False, help="Skip until to process",
```
Comment (mlissner) [inline cl/corpus_importer/management/commands/import_tn.py:181]: Make this an int, and make it default to 0 or -1. Should simplify things above and it'll provide better argparse'ing since non-ints won't be allowed.

File: cl/corpus_importer/management/commands/import_tn.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+class Command(VerboseCommand):
+    help = "Import TN data corpus received from TN Workers Comp boards."
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            "--input-dir",
+            help="The directory containing our data.",
+            required=True,
+        )
```
Comment (mlissner) [inline cl/corpus_importer/management/commands/import_tn.py:173]: Argparse has a native type for file parameters. Dig it up and use it instead? It'll do nice things like checking permissions, etc.

File: cl/corpus_importer/management/commands/import_tn.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    logging.info("Starting import")
+    filepath = "%s/data.json" % dir
+    tn_corpus = add_neutral_citations(filepath)
+    if not ready:
+        case = [x for x in tn_corpus if x["label"] == int(skip_until)][0]
+        logging.info(
+            "Skipping until case %s labeled: %s", case["title"], case["label"]
+        )
+
+    for case in sorted(tn_corpus, key=lambda x: x["label"]):
```
Comment (mlissner) [inline cl/corpus_importer/management/commands/import_tn.py:116]: Seems like we should have the data sorted by the function I suggested above, before we get here.

File: cl/corpus_importer/management/commands/import_tn.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+            continue
+        logging.info(
+            "Processing label:%s for case:%s", case["label"], case["title"]
+        )
+        pdf_path = [
+            x
+            for x in glob("%s/%s/*.pdf" % (dir, case["label"]))
+            if "stamped" not in x
+        ][0]
+        pdf_data = open(pdf_path, "r").read()
```
Comment (mlissner) [inline cl/corpus_importer/management/commands/import_tn.py:129]: You don't close this file. Use a context manager?

File: cl/corpus_importer/management/commands/import_tn.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+            ready = True
+        if not ready:
+            continue
+        logging.info(
+            "Processing label:%s for case:%s", case["label"], case["title"]
+        )
+        pdf_path = [
+            x
+            for x in glob("%s/%s/*.pdf" % (dir, case["label"]))
+            if "stamped" not in x
```
Comment (mlissner) [inline cl/corpus_importer/management/commands/import_tn.py:127]: Seems like you could do `%s/%s/*stamped*.pdf` or similar to simplify this logic? (using `glob` and an if function and a loop to filter seems like a lot.)

File: cl/corpus_importer/management/commands/import_tn.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+            logging.warn(
+                "Document already in database. See: %s at %s"
+                % (op.get_absolute_url(), op.cluster.case_name)
+            )
+
+        docket, opinion, cluster, citations, error = make_objects(
+            make_item(case),
+            Court.objects.get(pk=case["court"]),
+            sha1(force_bytes(pdf_data)),
+            open(pdf_path, "r").read(),
```
Comment (mlissner) [inline cl/corpus_importer/management/commands/import_tn.py:144]: Oof. These three lines all repeat something you already did. You get the Court each and every time instead of caching it, you compute the sha1 a second time, and you open and read the file (without closing it) a second time. Refactor please?

File: cl/corpus_importer/management/commands/import_tn.py (hunk 3 above)
Comment (flooie) [reply cl/corpus_importer/management/commands/import_tn.py:41]: Well - they may have retired prior - but the old guys last case was 2019 and the new guys first case was in 2020

File: cl/corpus_importer/management/commands/import_tn.py (hunk 9 above)
Comment (flooie) [reply cl/corpus_importer/management/commands/import_tn.py:127]: I'll just clean the data before I send it - and remove this section altogether

File: cl/corpus_importer/management/commands/import_tn.py (hunk 1 above)
Comment (flooie) [reply cl/corpus_importer/management/commands/import_tn.py:91]: just yanked the whole function.  No need to dynamically generate what I can just add to our data set.

File: cl/corpus_importer/management/commands/import_tn.py (hunk 4 above)
Comment (flooie) [reply cl/corpus_importer/management/commands/import_tn.py:37]: √

File: cl/corpus_importer/management/commands/import_tn.py (hunk 5 above)
Comment (flooie) [reply cl/corpus_importer/management/commands/import_tn.py:181]: √

File: cl/corpus_importer/management/commands/import_tn.py (hunk 7 above)
Comment (flooie) [reply cl/corpus_importer/management/commands/import_tn.py:116]: √

File: cl/corpus_importer/management/commands/import_tn.py (hunk 8 above)
Comment (flooie) [reply cl/corpus_importer/management/commands/import_tn.py:129]: data refinement let me remove this

File: cl/corpus_importer/management/commands/import_tn.py (hunk 6 above)
Comment (flooie) [reply cl/corpus_importer/management/commands/import_tn.py:173]: Switched it to a file type.  I didn't see anything with directory specific types but I think the change to file input makes more sense anyway

File: cl/corpus_importer/management/commands/import_tn.py (hunk 10 above)
Comment (flooie) [reply cl/corpus_importer/management/commands/import_tn.py:144]: I see two of them but not the court object.  I look forward to your next comment on it

File: cl/corpus_importer/management/commands/import_tn.py (hunk 2 above)
Comment (flooie) [reply cl/corpus_importer/management/commands/import_tn.py:29]: Yeah, I like title case but knowing the data I knew there were no edge cases here.

File: cl/corpus_importer/tests.py [hunk 11]
```diff
@@ -476,6 +478,62 @@ def test_num_queries_ok(self):
...
+        )
+
+    @mock.patch(
+        "cl.corpus_importer.management.commands.import_tn.add_neutral_citations",
+        side_effect=[
+            json.loads(
+                open(
+                    os.path.join(test_dir, "tn_corpus_test_asset.json"), "r"
+                ).read()
+            )
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/corpus_importer/tests.py:522]: √

File: cl/corpus_importer/management/commands/import_tn.py (hunk 10 above)
Comment (mlissner) [reply cl/corpus_importer/management/commands/import_tn.py:144]: Yeah, that one will get cached by the DB, so it's not the end of the world, but remember that every time you do Court.objects.get, you send data over the network to the DB, have it look something up and then return it back to you over the network. Instead of doing that thousands of times, you could do it once at the beginning of the script and have the court data ready. 

Something like this near the beginning:

```
courts = {"tennct": Court.objects.get(pk="tennct"), "tennworkcp": Court.objects.get(pk="tennworkcp")}
```

Then on this line you could do:

```
courts[case["court"]]
```

And you'd save the lookup.

File: cl/corpus_importer/management/commands/import_tn.py (hunk 6 above)
Comment (mlissner) [reply cl/corpus_importer/management/commands/import_tn.py:173]: We have a helper util in argtypes_helpers.py or something if you really want to do a directory, but I figured you'd prefer the filetype one anyway.

File: cl/corpus_importer/tests.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (flooie) [reply cl/corpus_importer/tests.py]: this is removed when we no longer have to handle neutral citations

File: cl/corpus_importer/management/commands/import_tn.py (hunk 10 above)
Comment (flooie) [reply cl/corpus_importer/management/commands/import_tn.py]: Gotcha.

Comment (flooie) [reply cl/corpus_importer/management/commands/import_tn.py]: fixed.

File: cl/corpus_importer/fixtures/tenn_court_fixture.json [hunk 12]
```diff
@@ -0,0 +1,1284 @@
+[
+  {
+    "model": "people_db.position",
```
Comment (mlissner) [inline cl/corpus_importer/fixtures/tenn_court_fixture.json:3]: I suspect we don't need the full fixture for tests. Can we slim it down considerably? The advantage of doing so is two-fold. One, it takes up less space in the repo. Two, when you're debugging later, it's easier to understand which parts matter.

File: cl/corpus_importer/management/commands/import_tn.py (hunk 3 above)
Comment (mlissner) [reply cl/corpus_importer/management/commands/import_tn.py:41]: Marking unresolved. You're still parsing the date twice.

File: cl/corpus_importer/management/commands/import_tn.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+            "--log",
+            action="store_true",
+            default=False,
+            help="Choose to view info log lines.",
+        )
+        parser.add_argument(
+            "--skip-until",
+            default=False,
+            help="Skip until to process",
+            type=int,
```
Comment (mlissner) [inline cl/corpus_importer/management/commands/import_tn.py:158]: Type `int`, default `False` is a bit weird. I'd default to zero or -1.

File: cl/corpus_importer/tests.py [hunk 14]
```diff
@@ -476,6 +478,48 @@ def test_num_queries_ok(self):
...
+    def tearDown(self):
+        Docket.objects.all().delete()
+
+    def test_import(self):
+        """Can we import two cases successfully"""
+        pre_install_count = OpinionCluster.objects.all().count()
+        filepath = os.path.join(
+            self.test_dir, "tenn_test_files", "tn_corpus_test_asset.json"
+        )
+        file = open(filepath)
```
Comment (mlissner) [inline cl/corpus_importer/tests.py:499]: Dangling file handle, again. Sorry if I missed it the first time. I'd suggest you pretend you never learned how to open files this way. The only reason I can think of doing it like this is to prevent indentation that'd be caused by a context handler.

File: cl/corpus_importer/tests.py [hunk 15]
```diff
@@ -476,6 +478,48 @@ def test_num_queries_ok(self):
...
+            post_install_count,
+            msg="Did not get two installs",
+        )
+
+    def test_panel_selection(self):
+        """Can we choose panelist correctly?"""
+        filepath = os.path.join(
+            self.test_dir, "tenn_test_files", "tn_corpus_test_asset.json"
+        )
+        file = open(filepath)
```
Comment (mlissner) [inline cl/corpus_importer/tests.py:513]: Another here.

File: cl/corpus_importer/management/commands/import_tn.py (hunk 13 above)
Comment (flooie) [reply cl/corpus_importer/management/commands/import_tn.py:158]: thinking about it now - id probably just leave it with no default.  let it roll as none if its not addressed.

File: cl/corpus_importer/fixtures/tenn_court_fixture.json (hunk 12 above)
Comment (flooie) [reply cl/corpus_importer/fixtures/tenn_court_fixture.json:3]: slimmed down

File: cl/corpus_importer/tests.py (hunk 15 above)
Comment (flooie) [reply cl/corpus_importer/tests.py]: Fixed

File: cl/corpus_importer/tests.py (hunk 14 above)
Comment (flooie) [reply cl/corpus_importer/tests.py]: Fixed

File: cl/corpus_importer/management/commands/import_tn.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+    first_name = judges.split(" ")[0].title()
+
+    lead_author = Person.objects.get(
+        positions__court_id=case["court"],
+        is_alias_of=None,
+        name_first=first_name,
+    )
+
+    # Four Panelists exist, one retired end of 2019 and the other joined.
+    pub_date_year = parser.parse(case["pub_date"]).date().year
```
Comment (mlissner) [inline cl/corpus_importer/management/commands/import_tn.py:38]:
```suggestion
    pub_date = parser.parse(case["pub_date"])
```

File: cl/corpus_importer/management/commands/import_tn.py [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+    if case["court"] == "tennworkcompapp":
+        exclude = "Marshall" if pub_date_year == 2020 else "Pele"
```
Comment (mlissner) [inline cl/corpus_importer/management/commands/import_tn.py:40]:
```suggestion
        exclude = "Marshall" if pub_date.date().year == 2020 else "Pele"
```

File: cl/corpus_importer/management/commands/import_tn.py [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+        ).exclude(name_first=exclude)
+        panelists = ", ".join([x.name_full for x in panelists_query])
+    else:
+        panelists = case["judge"]
+
+    return {
+        "source": Docket.DIRECT_INPUT,
+        "cluster_source": "D",
+        "case_names": case["title"],
+        "case_dates": parser.parse(case["pub_date"]),
```
Comment (mlissner) [inline cl/corpus_importer/management/commands/import_tn.py:52]:
```suggestion
        "case_dates": pub_date,
```

File: cl/corpus_importer/management/commands/import_tn.py (hunk 3 above)
Comment (flooie) [reply cl/corpus_importer/management/commands/import_tn.py]: okay.  I believe resolved
